src/PowerShellEditorServices.Protocol/LanguageServer/EditorCommands.cs

```diff
@@ -38,6 +38,8 @@ public static readonly
 
     public class ClientEditorContext
     {
+        public string CurrentFileContent { get; set; }
```

Comment:
Is it possible to also send the current language?

We may want to also add that as a property to `ScriptFile`, and possibly skip analysis/parsing if the language isn't PowerShell.

Reply:
I've dug into this.

One thing I'm concerned about with this is that the language comes from the client and every client might set language ids differently. For example, in vscode, it's called `languageId` and it's set to `powershell` but in some editors it might be `ps` or whatever.

Also, vscode doesn't send us the languageId for other actions so the only thing that would be populating a `CurrentFileLanguage` would be from our doing.

Additionally, I noticed that analysis works fine in PowerShell untitled files but doesn't work in PlainText ones even though this change creates ScriptFiles for the PlainText ones. That means we're doing _some_ sort of skipping of analysis if it's not actually a powershell file.

If anything, I think this change could be a boolean `IsPowerShellScript` or `IsPowerShell` but since this is not the only thing that creates `ScriptFile`s, I'm hesitant.

Do you have any additional knowledge here?

Reply:
> One thing I'm concerned about with this is that the language comes from the client and every client might set language ids differently. For example, in vscode, it's called languageId and it's set to powershell but in some editors it might be ps or whatever.

> Also, vscode doesn't send us the languageId for other actions so the only thing that would be populating a CurrentFileLanguage would be from our doing.

All good points, probably best to skip that, at least until we can come up with a consistent way.

> Additionally, I noticed that analysis works fine in PowerShell untitled files but doesn't work in PlainText ones even though this change creates ScriptFiles for the PlainText ones. That means we're doing some sort of skipping of analysis if it's not actually a powershell file.

Yeah on second thought this makes sense.  Analysis only gets triggered when the client sends us a `didOpen`/`didChange` notifcation.  That will never be the case for non-powershell files.  Still wise to skip parsing imo.

> If anything, I think this change could be a boolean IsPowerShellScript or IsPowerShell

This is the way to go I think

> but since this is not the only thing that creates ScriptFiles, I'm hesitant.

Here's what I'd suggest:

1. Add a new property `IsPowerShell` to `ScriptFile`
1. Add a second constructor that takes a bool for `isPowerShell` that sets the value (make it internal)
1. In the existing contructor(s) set `IsPowerShell` to `true`
1. Change `ParseFileContents` to return early if `IsPowerShell` is `false`

Reply:
Every approach I take for this leaves me in some awkward spot.

I attempted to implement what you asked, but the problem I ran into there is that `GetFile` and `GetFileBuffer` can't really tell whether or not to pass in `true` or `false` for `IsPowerShell` into the ScriptFile constructor.

I suppose `GetFile` _could_ check the file extension... which is ok I guess...

But what about `GetFileBuffer` which handles Untitled files?

I then thought to add an `isPowerShell` parameter to `GetFileBuffer` (which is kinda hacky IMO) and asked myself... ok how does `ConvertClientEditorContext` which calls `GetFileBuffer` know the language of the Untitled file? We'd need to be able to have the client pass whether or not the language is PowerShell.

Then we're back at the previous problem. We can't depend (yet) on what the client calls their file types.

However... `ConvertClientEditorContext` might be able to _always_ pass in `false` for `isPowerShell` in `GetFileBuffer` because when calling `$psEditor.GetEditorContext()` on any file, the PowerShell files should already have `ScriptFile`s associated with them because they get created when `HandleDidOpenTextDocumentNotification` gets fired.

Gah, I just thought of another thing... what if someone _changes_ the language mode of a file?  We'd need to be able to update the `IsPowerShell` value so that parsing can happen (or not). Pretty sure LSP doesn't give us a "Change Language Mode" event.

This... is pretty extensive. What do you think? I'm inclined to say this is out of scope for this PR and that I can open an issue for a perf gain.

Reply:
I think it's fair to call this out of scope.

We should probably have it at some point I suppose, but if we don't need it now, then I'm happy as is.

It sounds like from a layering perspective, the client needs to handle the "filetype" and we should just have a small list of known filetypes or languages we support and have a default case for the rest.

The for any particular client, it is responsible for translating the client's native filetype naming system to ours. (Since people are already implementing a client to interact with us anyway, this is just another protocol feature).

That might not be possible with the LSP and how VSCode uses it. But (1) it's not a problem we need to solve right now and (2) if the LSP is supposed to be editor agnostic and doesn't have a generalised notion of a filetype/language, then that's an issue for them I imagine.

Reply:
The performance gain is an added bonus, my main concern though is the change to `$psEditor.GetEditorContext()` behavior without a way to mitigate it. Right now, as a consumer of that API, I can be fairly certain that the context I get is PowerShell.  Granted that's because it breaks if it isn't, but still 😉 

> We'd need to be able to have the client pass whether or not the language is PowerShell.
> Then we're back at the previous problem. We can't depend (yet) on what the client calls their file types.

I don't think this is really a problem for this particular API.  This request isn't part of the LSP, so if the client actually responds then it has explicitly implemented *our* request.  If they're doing that they can almost definitely tell us if the current file is PowerShell.  And if not they can always just say it is.

I'm fine with calling this out of scope for this PR, but I'd really like this implemented before it gets released.  I'll take it if needed.

Reply:
Well if we want to fix this before a release let's just figure this out in this PR 😊

Here's what I've got so far:

* the client will send an `IsPowerShell` in the request

I'll then add the ability to:
* set `IsPowerShell` in the `GetFile*` (thinking about renaming to/adding a `GetOrCreateScriptFile` for this... Thoughts) 
* in the `HandleDidOpen` we either:
    * don't use `GetFile` and just always recreate the `ScriptFile`
    * give `GetFile` ( or `GetOrCreateScriptFile`) a force parameter that will recreate the `ScriptFile`

This second bullet is to handle the scenario when someone goes from a plaintext Untitled file, to a PowerShell Untitled file (which I personally do often enough).

I'm not sure what the frequency of `OnDidOpen` as in, if it fires every time you open a file (close, open, close, open) or if it happens only the first time you open it. I would guess the former... Which means we're recreating the `ScriptFile` over and over... Maybe we get do the `GetFile` and if IsPowerShell is false, then we can do the `GetFile` with `Force`. 

Typed this up on my phone so sorry for the formatting :/

Thoughts?